compiler/rustc_hir_analysis/src/coherence/builtin.rs

```diff
@@ -37,22 +37,23 @@ pub(super) fn check_trait<'tcx>(
 ) -> Result<(), ErrorGuaranteed> {
     let lang_items = tcx.lang_items();
     let checker = Checker { tcx, trait_def_id, impl_def_id, impl_header };
-    let mut res = checker.check(lang_items.drop_trait(), visit_implementation_of_drop);
-    res = res.and(checker.check(lang_items.copy_trait(), visit_implementation_of_copy));
-    res = res.and(checker.check(lang_items.const_param_ty_trait(), |checker| {
-        visit_implementation_of_const_param_ty(checker, LangItem::ConstParamTy)
-    }));
-    res = res.and(checker.check(lang_items.unsized_const_param_ty_trait(), |checker| {
-        visit_implementation_of_const_param_ty(checker, LangItem::UnsizedConstParamTy)
-    }));
-
-    res = res.and(
-        checker.check(lang_items.coerce_unsized_trait(), visit_implementation_of_coerce_unsized),
-    );
-    res.and(
-        checker
-            .check(lang_items.dispatch_from_dyn_trait(), visit_implementation_of_dispatch_from_dyn),
-    )
+    (|| {
+        checker.check(lang_items.drop_trait(), visit_implementation_of_drop)?;
+        checker.check(lang_items.copy_trait(), visit_implementation_of_copy)?;
+        checker.check(lang_items.const_param_ty_trait(), |checker| {
+            visit_implementation_of_const_param_ty(checker, LangItem::ConstParamTy)
+        })?;
+        checker.check(lang_items.unsized_const_param_ty_trait(), |checker| {
+            visit_implementation_of_const_param_ty(checker, LangItem::UnsizedConstParamTy)
+        })?;
+        checker.check(lang_items.coerce_unsized_trait(), visit_implementation_of_coerce_unsized)?;
+        checker.check(
+            lang_items.dispatch_from_dyn_trait(),
+            visit_implementation_of_dispatch_from_dyn,
+        )?;
+        checker.check(lang_items.pointer_like(), visit_implementation_of_pointer_like)?;
+        Ok(())
+    })()
```

Comment:
why do we need to use a closure here?

Reply:
It's like a try block. I found the repeated `res = res.and()` to be incredibly distracting.

Reply:
but the function itself already returns `Result<(), ErrorGuaranteed>`, why does removing this closure not work. I.e. we should be able to continue using `?` to eagerly return from the fn

Reply:
hahahahahahaha